backend/homework/4/questions/src/main/java/question1/StudentUtil.java

```diff
@@ -0,0 +1,67 @@
+package question1;
+
+import com.ganesh.LogMaster;
+import java.util.HashMap;
+
+public class StudentUtil {
+    public static double[] calculateGPA(int[] studentIdList, char[][] studentsGrades) throws IllegalArgumentException, MissingGradeException{
+        if (studentsGrades.length != studentIdList.length){
+            throw new IllegalArgumentException("studentIdList & studentsGrades are out-of-sync. studentIdList.length: " + studentIdList.length + ", studentsGrades.length: " + studentsGrades.length);
+        }
+        //make hashmap to eliminate multiple if else later
+        HashMap<Character, Double> gradeMapping = new HashMap<>();
+        gradeMapping.put('A', 4.0);
+        gradeMapping.put('B', 3.0);
+        gradeMapping.put('C', 2.0);
+        double[] finalResults = new double[studentIdList.length];
+        for (int index = 0; index < studentIdList.length; index++) {
+            char[] grades = studentsGrades[index];
+            int length = grades.length;
+            double gpa = 0.0;
+            //find sum of all grades
+            for (char grade : grades) {
+                if (grade == ' '){
+                    throw  new MissingGradeException(studentIdList[index]);
+                }
+                gpa += gradeMapping.get(grade);
+            }
+            gpa /= length;
+            finalResults[index] = gpa;
+        }
+        return finalResults;
+    }
+
+    public static int[] getStudentsByGPA(double lower, double higher, int[] studentIdList, char[][] studentsGrades) {
+        //data validity check
+        if (lower < 0 || higher > 4) {
```

Comment:
Lets move this hardcoded values in a constant variable and shift into a separate file.